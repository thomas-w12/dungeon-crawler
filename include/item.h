#ifndef ITEM_H
#define ITEM_H

#define ITEM_TYPES (const char*[]){"sword", "stone", "shield", "ladder"}
#define ITEM_DESCRIPTIONS (const char*[]){"Fight off monsters", "Stone the giant", "Shield against trap", "Climb out of trap"}

typedef struct Item{
    // Item type and description depends on the item id
    int ID;
    char type[10];
    char description[20];
} Item;

Item* Item_construct(int ID);
void displayItem(Item* item);
void displayItems(Item* items[], int itemCount);
void freeItem(Item* item);
void freeItems(Item* items[], int* itemsCount);
// void addItemToRoom(Room* room, Item item);
// void addItemToPlayer(Player* player, Item item);
// void removeItemFromRoom(Room* room, Item item); // Remove item after player collects item
// void removeItemFromPlayer(Player* player, Item item); // Remove item after player uses them or dies

#endif