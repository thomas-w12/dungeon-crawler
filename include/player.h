#ifndef PLAYER_H
#define PLAYER_H

#include <stdio.h>
#include "item.h"
#include "room.h"
// struct Room;

#define MAX_PLAYER_ITEMS 3

typedef struct Player {
    char name[10]; // Tell player to enter name at begining. Max length of about 10 characters
    int currentRoom; //ID of current room
    int health; // Health bar
    int score;
    Item* inventory[MAX_PLAYER_ITEMS]; // List of items the player has
    int itemsCount;
} Player;

void moveNorth(Player* player, Room* currentRoom);
void moveSouth(Player* player, Room* currentRoom);
void moveWest(Player* player, Room* currentRoom);
void moveEast(Player* player, Room* currentRoom);
void pickUpItem(Player* player, Item* item);

#endif