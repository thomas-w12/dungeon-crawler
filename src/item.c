#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "../include/item.h"
#include "../include/global.h"


Item* Item_construct(int ID) {
    Item* newItem = malloc(sizeof(Item));
    if (!newItem) {
        printf("Error: Could not allocate memory for Item.\n");
        return NULL;
    }
    newItem->ID = ID;
    strcpy(newItem->type, ITEM_TYPES[ID]);
    strcpy(newItem->description, ITEM_DESCRIPTIONS[ID]);
    return newItem;
}

ItemNode* ItemNode_construct(Item* item){
    ItemNode* itemNode = malloc(sizeof(ItemNode));
    if (itemNode == NULL){
        perror("Could not allocate memory for ItemNode");
        return NULL;
    }

    itemNode->data = item;
    itemNode->next = NULL;

    return itemNode;
}

ItemNode* ItemList_insert(ItemNode** head, Item* item){
    ItemNode* itemNode = ItemNode_construct(item);
    if (itemNode == NULL){
        perror("Failed Item List insert");
        return NULL;
    }
    // cannot have duplicate events
    if (*head == NULL){
        *head = itemNode;
    }else{
        ItemNode* curr = *head;

        while(curr->next != NULL){
            if (curr->data->ID == item->ID){
                perror("Duplicate items are not allowed");
                return NULL;
            }
            curr = curr->next;
        }

        curr->next = itemNode;
    }
    return itemNode;
}

Item* ItemList_deleteAtIndex(ItemNode** head, int itemIndex){
    Item* item = NULL;

    if (*head == NULL){
        return item;
    }else if (itemIndex == 0){// remove head
        if ((*head)->next == NULL){
            ItemNode* temp = (*head);
            item = (*head)->data;
            *head = NULL;
            free(temp);
            return item;
        }
        ItemNode* temp = (*head)->next;
        item = (*head)->data;
        free(*head);
        *head = temp;
    }else{
        ItemNode* prev = *head;
        ItemNode* curr = (*head)->next;

        int count = 1; //because we are not staring from the beginnig
        while((curr != NULL) && (itemIndex > count)){
            prev = curr;
            curr = curr->next;
            count++;
        }
        
        if (curr != NULL){
            prev->next = curr->next;
            item = curr->data;
            free(curr);
        }
        displayItem(item);
    }
    return item;
}

Item* ItemList_deleteItemID(ItemNode** head, int itemID){
    Item* item = NULL;

    if (*head == NULL){
        return item;
    }else if (itemID == (*head)->data->ID){// remove head
        if ((*head)->next == NULL){
            ItemNode* temp = (*head);
            item = (*head)->data;
            *head = NULL;
            free(temp);
            return item;
        }
        ItemNode* temp = (*head)->next;
        item = (*head)->data;
        free(*head);
        *head = temp;
    }else{
        ItemNode* prev = *head;
        ItemNode* curr = (*head)->next;

        int count = 1; //because we are not staring from the beginnig
        while((curr != NULL) && (itemID != curr->data->ID)){
            prev = curr;
            curr = curr->next;
            count++;
        }
        
        if (curr != NULL){
            prev->next = curr->next;
            item = curr->data;
            free(curr);
        }
    }
    return item;
}

int itemListCount(ItemNode* head){
    ItemNode* curr = head;
    int count = 0;
    while (curr != NULL){
        curr = curr->next;
        count++;
    }
    return count;
}

void itemListToArr(ItemNode* head, Item* arr[], int arrSize){
    ItemNode* curr = head;
    for (int i=0; i<arrSize; i++){
        if (curr == NULL){
            arr[i] = NULL;
        }else{
            arr[i] = curr->data;
            curr = curr->next;
        }
    }
}

void displayItem(Item* item){
    if (item != NULL) {
        printf("\nItem - ID: %d\tType: %s\tDescription: %s", item->ID, item->type, item->description);
    }else{
        printf("Item is NULL");
    }
}

void displayItems(Item* items[], int itemCount, int maxItems){
    if (itemCount == 0){
        printf("There are no items!");
        return;
    }

    for (int i=0; i<maxItems; i++){
        displayItem(items[i]);
    }
}

void printItemList(ItemNode* head){
    if (head == NULL){
        printf("There are no items!");
        return;
    }
    ItemNode* curr = head;
    while (curr != NULL){
        // printf("%d->", curr->data);
        displayItem(curr->data);
        curr = curr->next;
    }
}

void freeItem(Item* item){
    if (item == NULL) {
        return;
    }
    free(item);
}

void freeItemNode(ItemNode* itemNode){
    if (itemNode == NULL) return;
    freeItemNode(itemNode->next);
    free(itemNode->data);
    free(itemNode);
}

void freeItemList(ItemNode** headPtr){
    if (*headPtr == NULL) return;
    freeItemNode(*headPtr);
    *headPtr = NULL;
}

void freeItems(Item* items[], int* itemsCount){
    for (int i=0; i<*itemsCount; i++){
        free(items[i]);
    }
    *itemsCount = 0;
}