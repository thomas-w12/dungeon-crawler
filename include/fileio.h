#ifndef FILEIO_H
#define FILEIO_H

#include <stdio.h>
#include "room.h"
#include "player.h"

#define DNE -1

int saveLayout(const char* layoutStateFPath, Room** rooms, int roomCount);
int loadLayout(const char* layoutStateFPath, Room** rooms, int* roomCount);
int savePlayerState(const char* playerStateFPath, Player* player);
int loadPlayerState(const char* playerStateFPath, Player* player);

#endif
