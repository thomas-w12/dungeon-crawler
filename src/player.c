#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "../include/room.h"
#include "../include/item.h"
#include "../include/player.h"


Player* Player_construct(char* name, int currentRoom, int health, int score, Room* currentRoomPtr){
    Player* player = (Player*)malloc(sizeof(Player));
    if (player == NULL){
        printf("\nCould not create player");
        return NULL;
    }

    player->name = strdup(name);
    if (player->name == NULL){
        printf("\nCould not create player name");
        free(player);
        return NULL;
    }
    player->currentRoom = currentRoom;
    player->health = health;
    player->score = score;
    
    player->inventory = NULL;
    player->currentRoomPtr = currentRoomPtr;

    return player;
}

void displayPlayer(Player* player){
    printf("\nPlayer - Name: %s\tCurrent Room: %d\tHealth: %d\tScore: %d", player->name, player->currentRoom, player->health, player->score);
    displayPlayerInventory(player);
}

void displayPlayerInventory(Player* player){
    printf("\nItems in inventory (%d): ", itemListCount(player->inventory));
    printItemList(player->inventory);
}

void updatePlayerRoom(Player* player, Room* currentRoom){
    player->currentRoom = currentRoom->ID;
    player->currentRoomPtr = currentRoom;
}

void pickUpItem(Player* player, int pickupItemID){
    if (itemListCount(player->inventory) >= MAX_PLAYER_ITEMS){
        printf("\nInventory is full. Cannot pick up item.");
        return;
    }

    // Find the item with matching ID in the current room
    Item* pickupTargetItem = ItemList_deleteItemID(&player->currentRoomPtr->items, pickupItemID);
    if (pickupTargetItem == NULL) {
        printf("\nInvalid item ID\n");
        return;
    }

    // Add item to player inventory
    ItemList_insert(&player->inventory, pickupTargetItem);
    printf("\nYou picked up an Item. Press i to see your inventory\n");
}

void dropItem(Player* player, int dropItemID){
    if (itemListCount(player->inventory) == 0){
        printf("\nInventory is empty. Cannot drop item.");
        return;
    }

    // Find the item with matching ID in the player's inventory
    Item* dropTargetItem = ItemList_deleteItemID(&player->inventory, dropItemID);

    if (dropTargetItem == NULL) {
        printf("\nInvalid item ID\n");
        return;
    }

    // Add item to room
    ItemList_insert(&player->currentRoomPtr->items, dropTargetItem);
    printf("\nYou dropped an Item. Press i to see your inventory\n");
}

void decreasePlayerHealth(Player* player, int damage){
    player->health -= damage;
}

void increasePlayerHealth(Player* player, int health){
    player->health += health;
}

void increasePlayerScore(Player* player, int score){
    player->score += score;
}

void freePlayer(Player* player){
    if (player == NULL){
        return;
    }
    free(player->name);
    freeItemList(&player->inventory);
    free(player);
}

void triggerEvent(Room* room, Player* player){
    EventNode* currEvent = room->events;
    while (currEvent != NULL){
        switch(currEvent->data){
            case NORMAL:
                printf("\nThis room is a normal room");
                // delete event
                EventList_delete(&room->events, currEvent->data);
                break;
            case BLOCKED:
                printf("\nThis room is blocked");
                int input;
                scanf(" %d", &input);
                break;
            case LOCKED:
                printf("\nThis room is locked, you need a key to open it");
                break;
            case TRAP:
                printf("\nYou just got caught in a bobby trap! Youre health decreased by 25%%");
                decreasePlayerHealth(player, 25);
                break;
            case PIT_TRAP:
                printf("\nYou just fell into a pit");
                break;
            case PUZZLE:
                printf("\nYou have to solve this puzzel to coninue to this room");
                break;
            case NPC_BOSS:
                printf("\nYou just met an NpC boss");
                printf("\nFighting...");
                printf("\nYou Lost 50%% health");
                break;
        }

        currEvent = currEvent->next;
    }
}