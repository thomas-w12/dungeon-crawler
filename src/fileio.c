#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../include/fileio.h"
#include "../include/event.h"

int saveLayout(const char* layoutStateFPath, Room** rooms, int roomCount) {
    FILE* file = fopen(layoutStateFPath, "w");

    if (file == NULL) {
        perror("Error opening layout file");
        return EXIT_FAILURE;
    }

    fprintf(file, "%d\n", roomCount);

    for (int i = 0; i < roomCount; i++) {
        Room* room = rooms[i];
        fprintf(file, "%d,%s,%s,", room->ID, room->name, room->description);

        fprintf(file, "%d,%d,%d,%d,", 
                (room->north ? room->north->ID : DNE),
                (room->south ? room->south->ID : DNE),
                (room->west ? room->west->ID : DNE),
                (room->east ? room->east->ID : DNE));

        int arr[MAX_ROOM_EVENTS];
        eventListToArray(room->events, arr, MAX_ROOM_EVENTS);
        for (int j = 0; j < MAX_ROOM_EVENTS; j++) {
            fprintf(file, "%d,", arr[j]);
        }
        
        for (int j = 0; j < room->itemsCount; j++) {
            fprintf(file, "%d,", room->items[j]->ID);
        }
        fprintf(file, "\n");
    }

    fclose(file);
    return EXIT_SUCCESS;
}

int loadLayout(const char* layoutStateFPath, Room** rooms, int* roomCount) {
    FILE* file = fopen(layoutStateFPath, "r");

    if (file == NULL) {
        perror("Error opening layout file");
        return EXIT_FAILURE;
    }

    fscanf(file, "%d", roomCount);

    for (int i = 0; i < *roomCount; i++) {
        rooms[i] = (Room*)malloc(sizeof(Room));
        if (rooms[i] == NULL) {
            printf("Memory allocation failed for room %d\n", i);
            continue;
        }

        int id, northID, southID, westID, eastID;
        char name[10], description[30];

        fscanf(file, "%d,%10[^,],%30[^,],%d,%d,%d,%d,", 
               &id, name, description, &northID, &southID, &westID, &eastID); //30 here reads at most 30 characters as the description and 10 as the name

        rooms[i]->ID = id;
        strcpy(rooms[i]->name, name);
        strcpy(rooms[i]->description, description);
        rooms[i]->north = (northID != DNE) ? rooms[northID] : NULL;
        rooms[i]->south = (southID != DNE) ? rooms[southID] : NULL;
        rooms[i]->west = (westID != DNE) ? rooms[westID] : NULL;
        rooms[i]->east = (eastID != DNE) ? rooms[eastID] : NULL;
        rooms[i]->itemsCount = 0;

        rooms[i]->events = NULL;
        for (int j=0; j<MAX_ROOM_EVENTS; j++){
            int event;
            fscanf(file, "%d,", &event);
            if (event != DNE){
                rooms[i]->events = EventList_insert(rooms[i]->events, event);
            }
        }

        while (fgetc(file) != '\n' && !feof(file)) {
            fseek(file, -1, SEEK_CUR); // Go back one char from current position(fgetc pushed it forward)
            int itemID;
            fscanf(file, "%d,", &itemID);
            rooms[i]->items[rooms[i]->itemsCount] = Item_construct(itemID);
            rooms[i]->itemsCount++;
        }
    }

    fclose(file);
    return EXIT_SUCCESS;
}

int savePlayerState(const char* playerStateFPath, Player* player) {
    FILE* file = fopen(playerStateFPath, "w");

    if (file == NULL) {
        perror("Error opening player state file");
        return EXIT_FAILURE;
    }

    fprintf(file, "%s,%d,%d,%d,", player->name, player->currentRoom, player->health, player->score);
    for (int i = 0; i < MAX_PLAYER_ITEMS; i++) {
        if (player->inventory[i] != NULL) {
            fprintf(file, "%d,", player->inventory[i]->ID);
        }
    }

    fprintf(file, "\n");
    fclose(file);
    return EXIT_SUCCESS;
}

int loadPlayerState(const char* playerStateFPath, Player* player) {
    FILE* file = fopen(playerStateFPath, "r");

    if (file == NULL) {
        perror("Error opening player state file");
        return EXIT_FAILURE;
    }

    fscanf(file, "%9[^,],%d,%d,%d,", player->name, &player->currentRoom, &player->health, &player->score);
    player->itemsCount = 0;

    while (fgetc(file) != '\n' && !feof(file)) {
        int itemID;
        fscanf(file, "%d,", &itemID);
        player->inventory[player->itemsCount++] = Item_construct(itemID);
    }

    fclose(file);
    return EXIT_SUCCESS;
}
