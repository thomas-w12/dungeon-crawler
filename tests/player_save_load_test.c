#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include "../include/player.h"
#include "../include/fileio.h"
#include "../include/item.h"

int main() {
    srand(time(0));

    // Create items to be added to the room
    Item* item1 = Item_construct(0);
    Item* item2 = Item_construct(1);
    Item* item3 = Item_construct(2);

    ItemNode* itemsHead = NULL;
    ItemList_insert(&itemsHead, item1);
    ItemList_insert(&itemsHead, item2);
    ItemList_insert(&itemsHead, item3);

    // create room
    Room* startingRoom = Room_construct(0, NULL, NULL, NULL, NULL, NULL, itemsHead);
    if (startingRoom == NULL) {
        printf("Error creating starting room.\n");
        return EXIT_FAILURE;
    }

    // Create a player
    Player* player = Player_construct("Hero", startingRoom->ID, 100, 50, startingRoom);

    // Pick up items from the room
    pickUpItem(player, 0); // Pick up item with ID 0
    pickUpItem(player, 1); // Pick up item with ID 1

    // Save player state to a file
    char playerStateFPath[] = "tests/saved_test_games/player_save_load_test_output.txt";
    int saveStatus = savePlayerState(playerStateFPath, player);
    if (saveStatus != EXIT_SUCCESS) {
        printf("Error saving player state to file.\n");
        return EXIT_FAILURE;
    }

    // Load player state from the file
    Player* loadedPlayer = Player_construct("", 0, 0, 0, NULL);
    int loadStatus = loadPlayerState(playerStateFPath, loadedPlayer);
    if (loadStatus != EXIT_SUCCESS) {
        printf("Error loading player state from file.\n");
        return EXIT_FAILURE;
    }

    // Compare the loaded player with the original player
    if (strcmp(player->name, loadedPlayer->name) != 0 ||
        player->currentRoom != loadedPlayer->currentRoom ||
        player->health != loadedPlayer->health ||
        player->score != loadedPlayer->score) {
        printf("Loaded player data does not match saved player data.\n");
        return EXIT_FAILURE;
    }

    // Compare the inventory
    ItemNode* originalItem = player->inventory;
    ItemNode* loadedItem = loadedPlayer->inventory;
    while (originalItem != NULL && loadedItem != NULL) {
        if (originalItem->data->ID != loadedItem->data->ID) {
            printf("Loaded inventory item does not match saved inventory item.\n");
            return EXIT_FAILURE;
        }
        originalItem = originalItem->next;
        loadedItem = loadedItem->next;
    }
    if (originalItem != NULL || loadedItem != NULL) {
        printf("Loaded inventory size does not match saved inventory size.\n");
        return EXIT_FAILURE;
    }

    printf("Player data matches between saved and loaded player.\n");

    // Compare the saved file with the expected file
    char expectedPlayerStateFPath[] = "tests/saved_test_games/player_save_load_test_expected.txt";
    FILE* savedFile = fopen(playerStateFPath, "r");
    FILE* expectedFile = fopen(expectedPlayerStateFPath, "r");

    if (savedFile == NULL || expectedFile == NULL) {
        printf("Error opening saved or expected file for comparison.\n");
        return EXIT_FAILURE;
    }

    char savedLine[256];
    char expectedLine[256];
    int lineNumber = 1;
    while (fgets(savedLine, sizeof(savedLine), savedFile) && fgets(expectedLine, sizeof(expectedLine), expectedFile)) {
        if (strcmp(savedLine, expectedLine) != 0) {
            printf("Mismatch at line %d:\n", lineNumber);
            printf("Saved:    %s", savedLine);
            printf("Expected: %s", expectedLine);
            fclose(savedFile);
            fclose(expectedFile);
            return EXIT_FAILURE;
        }
        lineNumber++;
    }

    fclose(savedFile);
    fclose(expectedFile);

    printf("Saved player file matches expected player file.\n");

    // Free memory
    freePlayer(player);
    freePlayer(loadedPlayer);
    free(startingRoom);

    printf("Test passed!\n");
    return EXIT_SUCCESS;
}