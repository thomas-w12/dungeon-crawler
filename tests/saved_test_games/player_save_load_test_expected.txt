Hero,0,100,50,0,1,-1,
